ir/src/ir.rs

```diff
@@ -11,6 +19,28 @@ pub struct Axis {
     pub hidden: bool,
 }
 
+pub type DesignSpaceLocation = HashMap<String, f32>;
+
+/// A variable definition of a single glyph.
+///
+/// Defined in at least once position in designspace. If defined in
+/// many, presumed to vary continuously between positions and required
+/// to have variation compatible structure.
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct Glyph {
+    pub name: String,
+    pub instances: Vec<GlyphInstance>,
```

Comment:
didn't we want to call this `sources` and have a map of DesignSpaceLocation to GlyphInstances?
https://github.com/googlefonts/oxidize/pull/37/commits/a3efa92f95b22e8bf1ee6d9a68e0d3b400111783

I prefer `sources` to `instances` here, the former aligns with designspace terminology and conveys that these are the input to the interpolation, not the output. Instances can be used as sources (by associating them with a DS location and passed as inputs to the variation model) but can also be the result of the interpolation, when instancing the designspace at a location that's different the the sources' location.

Also I prefer if the DS location is the key and is not embedded in the GlyphInstance struct; that would ensure sources have unique locations; two sources with the same location make no sense and should be an error.

Reply:
sgtm

Reply:
The mapping seems to reflect what we want quite well, perhaps the thing to do is make DesignSpaceLocation a type that is hashable? - I believe the main thing you'd have to do is quantize the f32

Reply:
> make DesignSpaceLocation a type that is hashable

you mean making `DesignSpaceLocation` an alias of `Vec[DesignSpaceDimension]` where the dimension is a struct containing an axis name and the associated value? Implementing `Hash` trait for `HashMap` itself is [non trivial](https://internals.rust-lang.org/t/implementing-hash-for-hashset-hashmap/3817) and beyond me at this point.

> quantize the f32

that would be lossy, no?... How about using an [`ordered_float::OrderedFloat`](https://docs.rs/ordered-float/latest/ordered_float/struct.OrderedFloat.html) instead (as suggested [here](https://internals.rust-lang.org/t/f32-f64-should-implement-hash/5436)) which, unlike f32/f64, does implements `Eq` and `Hash` and can thus be used as key in hash maps/sets?

https://docs.rs/ordered-float/latest/ordered_float/struct.OrderedFloat.html

AFAIU, the only difference with built-in floats is that 

> NaN is sorted as greater than all other values and equal to itself, in contradiction with the IEEE standard

but NaNs are explicitly disallowed in UFOs, see https://github.com/unified-font-object/ufo-spec/pull/203

Reply:
Quantizing only for hashing wouldn't destroy the original value. https://crates.io/crates/ordered-float also sounds fine to me.

HOI support strikes me as something we should think about separately. Maybe we should file an issue?

Even under HOI I don't think it makes sense to have a Glyph with multiple definitions for the same location in design space. To my understanding HOI is about how you compute the values in between instances, not altering the actual instances. In abstract you could imagine an additional field defining how to interpolate, for napkin math purposes `how_to_interpolate: Map<(Location, Location), Curve>`.

Reply:
ok I am using ordered-float in https://github.com/googlefonts/fontmake-rs/pull/4/commits/498878e6de2c67be0eb70c4c6d9cf6a80d9e4bd8

I made DesignSpaceLocation a BTreeMap which already implements Hash and has a similar interface as HashMap but is ordered; I don't know how to implement Hash for HashMap ([this thread makes it look tricky](https://internals.rust-lang.org/t/implementing-hash-for-hashset-hashmap/3817)).

And I moved the location field from inside GlyphInstance struct to the key of the Glyph::sources hashmap, and also added a check that it be unique before inserting a new one.

Regarding HOI, 

> Even under HOI I don't think it makes sense to have a Glyph with multiple definitions for the same location in design space.

I didn't mean having multiple glyph sources at the same location; I mean a given location containing multiple axes with the same tag, which would be impossible if the location map was keyed on the axis tag. But since a designspace location is keyed on the axis _name_ (which by definition must be unique as it identifies a given designspace axis element) and not its _tag_, it's in theory possible for a location to refer to multiple axes with distinct names (e.g. "Weight#1", "Weight#2") but that are associated with the same tag ("wght"). So I will keep DesignSpaceLocation as a map instead of making it a vector of dimensions, which is also in line with the designspaceLib doc's about location: "Encodes a dictionary of { axisname: axisvalue }".

Reply:
> made DesignSpaceLocation a BTreeMap which already implements Hash

SGTM. That may explain why I saw BTreeMap in a few places where I didn't find it obvious why HashMap wouldn't be used.

Reply:
> I will keep DesignSpaceLocation as a map

+1, I think that is more aligned with what a glyph instance is: what the glyph looks like at one specific position in design space. HOI defines how to move non-linearly between those points. At a glance it doesn't seem readily expressible in the designspace/ufo format at all...?